### DISCUSSION TITLE
Restructuring of parameters associated with components

### DISCUSSION THREAD

**Discussion by abensonca** [galacticusorg/galacticus#333] [Announcements]
The latest revsion, [ce515f9](https://github.com/galacticusorg/galacticus/commit/ce515f9e4525c3464161e5183cbd7bfdc56c4e0d), makes a significant change to the structure of parameter files - specifically where parameters directly used by "components" should be placed.

For example, if your parameter file previously selected the `standard` black hole component, and set the black hole see mass as follows:
```
<componentBlackHole value="standard"/>
<blackHoleSeedMass value="100.0"/>
```
it would need to change to the following to work with the latest revision:
```
  <componentBlackHole value="standard">
    <massSeed value="100.0"/>
  </componentBlackHole>
```
i.e. the seed mass parameter is now a sub-parameter, and must appear inside the `componentBlackHole` element. Also, the names of many such parameters have been altered (as in this example) for improved consistency.

This is part of a longer term plan to refactor the way in which the components are structured within Galacticus to make the system more flexible.

There is an automated tool which can update your parameter file to work with this latest version of the code. For example, if you have a parameter file called `myParameters.xml` you can update it using:
```
./scripts/aux/parametersMigrate.pl myParameters.xml myParametersNew.xml
```
which will create the file `myParametersNew.xml` with the relevant parameters renamed and relocated.  This script is somewhat experimental (it was recently overhauled to make use of `git` history to help in figuring out which parameters need to be updated). So, if you run into any problems using it, let us know.

**Comment by MarcoMarinucci** [comment]
Hi, I have been using the parsMigrate on the baryonicPhysicsConstrained.xml file and it worked (the only change occurred at the outputTimes part). Now I'm trying to launch the new file, but I have an error that I attach:
Fatal error:
unmatched properties [Lyc : rest :  0.0000]
1 of 68 : SDSS_g : observed : 0.020
2 of 68 : SDSS_r : observed : 0.020
3 of 68 : SDSS_r : observed : 0.020
4 of 68 : SDSS_g : observed : 0.020
5 of 68 : Lyc : rest : 0.020
6 of 68 : HeliumContinuum : rest : 0.020
7 of 68 : OxygenContinuum : rest : 0.020
8 of 68 : SDSS_r : observed : 0.040
9 of 68 : SDSS_g : observed : 0.040
10 of 68 : Lyc : rest : 0.040
11 of 68 : SDSS_g : observed : 0.040
12 of 68 : SDSS_r : observed : 0.040
13 of 68 : HeliumContinuum : rest : 0.040
14 of 68 : OxygenContinuum : rest : 0.040
15 of 68 : SDSS_g : observed : 0.060
16 of 68 : SDSS_r : observed : 0.060
17 of 68 : Lyc : rest : 0.060
18 of 68 : SDSS_g : observed : 0.060
19 of 68 : SDSS_r : observed : 0.060
20 of 68 : HeliumContinuum : rest : 0.060
21 of 68 : OxygenContinuum : rest : 0.060
22 of 68 : SDSS_r : observed : 0.080
23 of 68 : Lyc : rest : 0.080
24 of 68 : SDSS_g : observed : 0.080
25 of 68 : SDSS_g : observed : 0.080
26 of 68 : HeliumContinuum : rest : 0.080
27 of 68 : SDSS_r : observed : 0.080
28 of 68 : OxygenContinuum : rest : 0.080
29 of 68 : OxygenContinuum : rest : 0.100
30 of 68 : SDSS_g : observed : 0.100
31 of 68 : SDSS_r : observed : 0.100
32 of 68 : Lyc : rest : 0.100
33 of 68 : SDSS_g : observed : 0.100
34 of 68 : HeliumContinuum : rest : 0.100
35 of 68 : SDSS_r : observed : 0.100
36 of 68 : SDSS_r : observed : 0.120
37 of 68 : SDSS_r : observed : 0.120
38 of 68 : SDSS_g : observed : 0.120
39 of 68 : SDSS_g : observed : 0.120
40 of 68 : SDSS_g : observed : 0.140
41 of 68 : SDSS_r : observed : 0.140
42 of 68 : SDSS_g : observed : 0.140
43 of 68 : SDSS_r : observed : 0.140
44 of 68 : SDSS_r : observed : 0.160
45 of 68 : SDSS_r : observed : 0.160
46 of 68 : SDSS_g : observed : 0.160
47 of 68 : SDSS_g : observed : 0.160
48 of 68 : SDSS_g : observed : 0.180
49 of 68 : SDSS_g : observed : 0.180
50 of 68 : SDSS_r : observed : 0.180
51 of 68 : SDSS_r : observed : 0.180
52 of 68 : SDSS_r : observed : 0.200
53 of 68 : SDSS_r : observed : 0.200
54 of 68 : SDSS_g : observed : 0.200
55 of 68 : SDSS_g : observed : 0.200
56 of 68 : SDSS_g : observed : 0.240
57 of 68 : SDSS_r : observed : 0.240
58 of 68 : SDSS_r : observed : 0.240
59 of 68 : SDSS_g : observed : 0.240
60 of 68 : HeliumContinuum : rest : 0.400
61 of 68 : Lyc : rest : 0.400
62 of 68 : OxygenContinuum : rest : 0.400
63 of 68 : OxygenContinuum : rest : 0.860
64 of 68 : Lyc : rest : 0.860
65 of 68 : HeliumContinuum : rest : 0.860
66 of 68 : HeliumContinuum : rest : 2.200
67 of 68 : Lyc : rest : 2.200
68 of 68 : OxygenContinuum : rest : 2.200
 Occurred at:
     function:Stellar_Luminosities_Index_From_Properties
       module:Stellar_Luminosities_Structure
         file:objects.stellar_luminosities.F90   [line 1149]
         
         

What is the problem? Are there other properties that need to be fixed?
Thanks

**Reply by abensonca** [reply]
Can you share the parameter file? It looks like you're trying to output luminosities, or are including some analysis that requires those luminosities (e.g. luminosity functions, or emission lines), but the luminosities are not specified in the parameter file.

**Reply by MarcoMarinucci** [reply]
I'm using the following (sorry for the txt format, but the xml is not supported).
[baryonicPhysicsConstrainedNew.txt](https://github.com/galacticusorg/galacticus/files/10215414/baryonicPhysicsConstrainedNew.txt)

Btw, I tried using the file galacticus/parameters/baryonicPhysicsConstrained.xml in the latest version and it doesn't work as well.

**Reply by abensonca** [reply]
It turns out that there were a couple of bugs exposed by this model. Those are being fixed in PR #345. I'll tag you in the comments there so you'll see when it's been merged into Galacticus - at that point you can update your copy to get the bug fixes. You'll also need to update your copy of the `datasets` repo as I had to make a fix there also.

Additionally, a couple of changes are needed to your parameter file. You had:
```
  <outputTimesMethod value="list">
    <times value="list">
    <redshifts value="0.00 0.02 0.04 0.06 0.08 0.10 0.12 0.14 0.16 0.18 0.20 0.24 0.28 0.32 0.70 0.78 0.86 0.94 1.02 2.20 2.50 2.75 3.00"/>\
  </times>
  </outputTimesMethod>
```
This should be:
```
  <outputTimes value="list">
    <redshifts value="0.00 0.02 0.04 0.06 0.08 0.10 0.12 0.14 0.16 0.18 0.20 0.24 0.28 0.32 0.70 0.78 0.86 0.94 1.02 2.20 2.50 2.75 3.00"/>
  </outputTimes>
```

The error you reported above arises because you included:
```
    <nodePropertyExtractor value="lmnstyEmssnLine">
      <lineNames value="balmerAlpha6565"/>
      <!--<depthOpticalISMCoefficient value="1.0"/>-->
    </nodePropertyExtractor>
```
This asks Galacticus to output that emission line luminosity at every output redshift. Doing so requires that the filters `Lyc`, `HeliumContinuum` and `OxygenContinuum` are computed for every redshift. But, in your original parameter file they were computed at only a subset of the redshifts. You can update the various `<luminosity*` parameters as follows to fix this:
```
  <!-- Luminosity calculations -->
  <luminosityFilter value="         SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g         SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r          SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g   SDSS_g           SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r   SDSS_r        emissionLineFilters    "/>
  
  <luminosityType value="         observed observed observed observed observed observed observed observed observed observed observed        observed observed observed observed observed observed observed observed observed observed observed         observed observed observed observed observed observed observed observed observed observed observed         observed observed observed observed observed observed observed observed observed observed observed rest       "/>
  
  <luminosityRedshift value="         0.02     0.04     0.06     0.08     0.10     0.12     0.14     0.16     0.18     0.20     0.24         0.02     0.04     0.06     0.08     0.10     0.12     0.14     0.16     0.18     0.20     0.24         0.02     0.04     0.06     0.08     0.10     0.12     0.14     0.16     0.18     0.20     0.24                 0.02     0.04     0.06     0.08     0.10     0.12     0.14     0.16     0.18     0.20     0.24        all    "/>

  
  <luminosityBandRedshift value="         0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10         0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10         0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10         0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10     0.10        all "/>
  <luminosityPostprocessSet value="         default  default  default  default  default  default  default  default  default  default  default         default  default  default  default  default  default  default  default  default  default  default         recent   recent   recent   recent   recent   recent   recent   recent   recent   recent   recent         recent   recent   recent   recent   recent   recent   recent   recent   recent   recent   recent default         "/>
```
In this I've used the special filter name `emissionLineFilters` which is automatically expanded to include all three ionizing continuum filters (`Lyc`, `HeliumContinuum` and `OxygenContinuum`), and I've used the `all` value for this filter in `luminosityRedshift` and `luminosityBandRedshift` - which is automatically expanded to include all redshifts listed in the `outputTimes` parameter.

With those changes and the bug fixes this model then runs successfully.

**Reply by abensonca** [reply]
The bug fixes are now merged, so you should be able to update your copy of Galacticus and try running your model again
